janitor/utils.py

```diff
@@ -2373,3 +2377,805 @@ def _sub_process_text_result_MultiIndex(index: pd.MultiIndex, result, df):
+
+    numeric_date_string = numeric_type, date_type, string_type
+    if any(numeric_date_string):
+        return None, None
+
+    raise ValueError(
+        """
+        conditional_join only supports
+        numeric, date, or string dtypes.
+        """
+    )
+
+
+# code copied from Stack Overflow
+# https://stackoverflow.com/a/47126435/7175713
+def _le_create_ranges(indices, len_right):
+    cum_length = len_right - indices
+    cum_length = cum_length.cumsum()
+    ids = np.ones(cum_length[-1], dtype=int)
+    ids[0] = indices[0]
+    ids[cum_length[:-1]] = indices[1:] - len_right + 1
+    return ids.cumsum()
+
+
+def _ge_create_ranges(indices):
+    cum_length = indices.cumsum()
+    ids = np.ones(cum_length[-1], dtype=int)
+    ids[0] = 0
+    ids[cum_length[:-1]] = -1 * indices[:-1] + 1
+    return ids.cumsum()
+
+
+def _equal_indices(left_c: pd.Series, right_c: pd.Series):
+    """
+    This uses binary search to get indices where
+    `left_c` is exactly equal to `right_c`.
+
+    Returns a tuple of (left_c, right_c)
+    """
+
+    # sorted data is a key requirement for binary search
+    right_group = right_c.groupby(right_c, sort=True)
+    # get the unique values and create a Pandas object
+    right_keys = [keys for keys, _ in right_group]
+    right_keys = pd.Index(right_keys)
+
+    # if there isn't any value from left_c in `right_keys`
+    # then there is no point in running the binary search
+    rows_to_keep = left_c.isin(right_keys)
+    if not rows_to_keep.any():
+        return None, None
+
+    # keep only values that definitely exist in `right_c`
+    left_c = left_c[rows_to_keep]
+    # get relative positions of `left_c` in
+    # the unique values from `right_keys`
+    search_indices = right_keys.searchsorted(left_c, side="left")
+
+    right_keys = right_keys.take(search_indices)
+
+    # keep only rows where `left` == `right`
+    rows_to_keep = left_c.array == right_keys
+
+    left_c = left_c.index[rows_to_keep]
+    right_keys = right_keys[rows_to_keep]
+
+    # get the index locations for the unique values in `right_c`
+    right_indices = right_group.indices
+    # map the indices to `right_keys`
+    # and subsequently concatenate to get
+    # the complete matched values from `right_c`
+    right_c = right_keys.map(right_indices)
+    right_c = np.concatenate(right_c)
+
+    # the counts of each unique value
+    # is used to `explode` left_c
+    # to match the number of rows from `right_c`
+    lengths = right_group.size()
+    lengths = right_keys.map(lengths)
+    left_c = left_c.repeat(lengths)
+
+    return left_c, right_c
+
+
+def _not_equal_indices(left_c: pd.Series, right_c: pd.Series):
+    """
+    This uses binary search to get indices where
+    `left_c` is exactly  not equal to `right_c`.
+    It is a combination of strictly less than
+    and strictly greater than indices.
+
+    Returns a tuple of (left_c, right_c)
+    """
```

Comment:
Minor grammatical nitpick.

```suggestion
    """
    Use binary search to get indices where
    `left_c` is exactly  not equal to `right_c`.
    It is a combination of strictly less than
    and strictly greater than indices.

    Returns a tuple of (left_c, right_c)
    """
```